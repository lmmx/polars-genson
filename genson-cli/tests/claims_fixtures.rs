use assert_cmd::Command;
use insta::{assert_snapshot, with_settings};
use std::fs;

/// Check if the current output matches the verified/blessed version
fn is_output_approved(snapshot_name: &str, output: &str) -> bool {
    let module_file = file!();
    let module_stem = std::path::Path::new(module_file)
        .file_stem()
        .unwrap()
        .to_string_lossy();
    let verified_path = format!("tests/verified/{}__{}.snap", module_stem, snapshot_name);

    if let Ok(verified_content) = fs::read_to_string(&verified_path) {
        if let Some(header_end) = verified_content.find("\n---\n") {
            let verified_output = &verified_content[header_end + 5..];
            return verified_output.trim() == output.trim();
        }
    }
    false
}

/// Run genson-cli with claims fixture from disk
fn run_genson_claims_fixture_from_disk(fixture_path: &str, name: &str, extra_args: &[&str]) {
    let mut cmd = Command::cargo_bin("genson-cli").unwrap();
    let mut args = vec![
        "--map-threshold",
        "0",
        "--unify-maps",
        "--wrap-root",
        "claims",
    ];
    // Add --ndjson flag if the fixture is a .jsonl file
    if fixture_path.ends_with(".jsonl") {
        args.push("--ndjson");
    }
    args.extend_from_slice(extra_args);
    args.push(fixture_path);
    let args_for_metadata = args.clone();
    cmd.args(args);

    let assert_output = cmd.assert().success();
    let output = assert_output.get_output();
    let stdout_str = String::from_utf8(output.stdout.clone()).unwrap();

    // Let stderr be visible if there's any debug output
    if !output.stderr.is_empty() {
        let stderr_str = String::from_utf8_lossy(&output.stderr);
        eprintln!("stderr from {}: {}", name, stderr_str);
    }

    let approved = is_output_approved(name, &stdout_str);

    with_settings!({
        info => &serde_json::json!({
            "approved": approved,
            "args": args_for_metadata[..args_for_metadata.len()-1], // exclude file path
            "fixture": fixture_path
        })
    }, {
        assert_snapshot!(name, stdout_str);
    });
}

#[test]
fn test_claims_fixture_l1_avro() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims_fixture_x4_L1.jsonl",
        "claims_fixture_l1__avro",
        &["--avro"],
    );
}

#[test]
fn test_claims_fixture_l1_jsonschema() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims_fixture_x4_L1.jsonl",
        "claims_fixture_l1__jsonschema",
        &[],
    );
}

#[test]
fn test_claims_fixture_l1_normalize() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims_fixture_x4_L1.jsonl",
        "claims_fixture_l1__normalize",
        &["--normalise"],
    );
}

#[test]
fn test_claims_fixture_l2_avro() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims_fixture_x4_L2.jsonl",
        "claims_fixture_l2__avro",
        &["--avro"],
    );
}

#[test]
fn test_claims_fixture_l2_jsonschema() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims_fixture_x4_L2.jsonl",
        "claims_fixture_l2__jsonschema",
        &[],
    );
}

#[test]
fn test_claims_fixture_l2_normalize() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims_fixture_x4_L2.jsonl",
        "claims_fixture_l2__normalize",
        &["--normalise"],
    );
}

#[test]
fn test_claims_fixture_l3_avro() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims_fixture_x4_L3.jsonl",
        "claims_fixture_l3__avro",
        &["--avro"],
    );
}

#[test]
fn test_claims_fixture_l3_jsonschema() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims_fixture_x4_L3.jsonl",
        "claims_fixture_l3__jsonschema",
        &[],
    );
}

#[test]
fn test_claims_fixture_l3_normalize() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims_fixture_x4_L3.jsonl",
        "claims_fixture_l3__normalize",
        &["--normalise"],
    );
}

#[test]
fn test_claims_fixture_l4_avro() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims_fixture_x4_L4.jsonl",
        "claims_fixture_l4__avro",
        &["--avro"],
    );
}

#[test]
fn test_claims_fixture_l4_jsonschema() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims_fixture_x4_L4.jsonl",
        "claims_fixture_l4__jsonschema",
        &[],
    );
}

#[test]
fn test_claims_fixture_l4_normalize() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims_fixture_x4_L4.jsonl",
        "claims_fixture_l4__normalize",
        &["--normalise"],
    );
}

// Following are numbered from the full x1818 fixture

#[test]
fn test_claims_fixture_x1818_l4_avro() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L4_MINIMAL.json",
        "claims_fixture_x1818_l4__avro",
        &["--avro"],
    );
}

#[test]
fn test_claims_fixture_x1818_l4_jsonschema() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L4_MINIMAL.json",
        "claims_fixture_x1818_l4__jsonschema",
        &[],
    );
}

#[test]
fn test_claims_fixture_x1818_l4_normalize() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L4_MINIMAL.json",
        "claims_fixture_x1818_l4__normalize",
        &["--normalise"],
    );
}

#[test]
fn test_claims_fixture_x1818_l5_avro() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L5_MINIMAL.json",
        "claims_fixture_x1818_l5__avro",
        &["--avro"],
    );
}

#[test]
fn test_claims_fixture_x1818_l5_jsonschema() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L5_MINIMAL.json",
        "claims_fixture_x1818_l5__jsonschema",
        &[],
    );
}

#[test]
fn test_claims_fixture_x1818_l5_normalize() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L5_MINIMAL.json",
        "claims_fixture_x1818_l5__normalize",
        &["--normalise"],
    );
}

#[test]
fn test_claims_fixture_x1818_l12_avro() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L12_MINIMAL.json",
        "claims_fixture_x1818_l12__avro",
        &["--avro"],
    );
}

#[test]
fn test_claims_fixture_x1818_l12_jsonschema() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L12_MINIMAL.json",
        "claims_fixture_x1818_l12__jsonschema",
        &[],
    );
}

#[test]
fn test_claims_fixture_x1818_l12_normalize() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L12_MINIMAL.json",
        "claims_fixture_x1818_l12__normalize",
        &["--normalise"],
    );
}

#[test]
fn test_claims_fixture_x1818_l14_avro() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L14_MINIMAL.json",
        "claims_fixture_x1818_l14__avro",
        &["--avro"],
    );
}

#[test]
fn test_claims_fixture_x1818_l14_jsonschema() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L14_MINIMAL.json",
        "claims_fixture_x1818_l14__jsonschema",
        &[],
    );
}

#[test]
fn test_claims_fixture_x1818_l14_normalize() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L14_MINIMAL.json",
        "claims_fixture_x1818_l14__normalize",
        &["--normalise"],
    );
}

#[test]
fn test_claims_fixture_x1818_l16_avro() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L16_MINIMAL.json",
        "claims_fixture_x1818_l16__avro",
        &["--avro"],
    );
}

#[test]
fn test_claims_fixture_x1818_l16_jsonschema() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L16_MINIMAL.json",
        "claims_fixture_x1818_l16__jsonschema",
        &[],
    );
}

#[test]
fn test_claims_fixture_x1818_l16_normalize() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L16_MINIMAL.json",
        "claims_fixture_x1818_l16__normalize",
        &["--normalise"],
    );
}

#[test]
fn test_claims_fixture_x1818_l26_avro() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L26_MINIMAL.json",
        "claims_fixture_x1818_l26__avro",
        &["--avro"],
    );
}

#[test]
fn test_claims_fixture_x1818_l26_jsonschema() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L26_MINIMAL.json",
        "claims_fixture_x1818_l26__jsonschema",
        &[],
    );
}

#[test]
fn test_claims_fixture_x1818_l26_normalize() {
    run_genson_claims_fixture_from_disk(
        "tests/data/claims/x1818_L26_MINIMAL.json",
        "claims_fixture_x1818_l26__normalize",
        &["--normalise"],
    );
}
