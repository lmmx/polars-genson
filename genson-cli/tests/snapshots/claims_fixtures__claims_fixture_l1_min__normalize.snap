---
source: genson-cli/tests/claims_fixtures.rs
expression: stdout_str
info:
  approved: false
  args:
    - "--map-threshold"
    - "0"
    - "--unify-maps"
    - "--wrap-root"
    - claims
    - "--ndjson"
    - "--normalise"
  fixture: tests/data/claims_fixture_x4_L1_min.jsonl
---
{"claims":{"P31":[{"mainsnak":{"property":"P31","datatype":"wikibase-item","datavalue":{"id":"Q15632617","datavalue__string":null,"text":null,"language":null,"time":null,"timezone":null,"before":null,"after":null,"precision":null,"calendarmodel":null}},"rank":"normal"}],"P40":[{"mainsnak":{"property":"P40","datatype":"wikibase-item","datavalue":{"id":"Q1049347","datavalue__string":null,"text":null,"language":null,"time":null,"timezone":null,"before":null,"after":null,"precision":null,"calendarmodel":null}},"rank":"normal"}],"P27":[{"mainsnak":{"property":"P27","datatype":"wikibase-item","datavalue":{"id":"Q30","datavalue__string":null,"text":null,"language":null,"time":null,"timezone":null,"before":null,"after":null,"precision":null,"calendarmodel":null}},"rank":"normal"}],"P345":[{"mainsnak":{"property":"P345","datatype":"external-id","datavalue":{"id":null,"datavalue__string":"ch0009881","text":null,"language":null,"time":null,"timezone":null,"before":null,"after":null,"precision":null,"calendarmodel":null}},"rank":"deprecated"}],"P108":[{"mainsnak":{"property":"P108","datatype":"wikibase-item","datavalue":{"id":"Q2944031","datavalue__string":null,"text":null,"language":null,"time":null,"timezone":null,"before":null,"after":null,"precision":null,"calendarmodel":null}},"rank":"normal"}],"P463":[{"mainsnak":{"property":"P463","datatype":"wikibase-item","datavalue":{"id":"Q209114","datavalue__string":null,"text":null,"language":null,"time":null,"timezone":null,"before":null,"after":null,"precision":null,"calendarmodel":null}},"rank":"normal"}],"P569":[{"mainsnak":{"property":"P569","datatype":"time","datavalue":{"id":null,"datavalue__string":null,"text":null,"language":null,"time":"+1966-02-18T00:00:00Z","timezone":0,"before":0,"after":0,"precision":11,"calendarmodel":"http://www.wikidata.org/entity/Q1985727"}},"rank":"normal"}],"P451":[{"mainsnak":{"property":"P451","datatype":"wikibase-item","datavalue":{"id":"Q284262","datavalue__string":null,"text":null,"language":null,"time":null,"timezone":null,"before":null,"after":null,"precision":null,"calendarmodel":null}},"rank":"normal"}],"P19":[{"mainsnak":{"property":"P19","datatype":"wikibase-item","datavalue":{"id":"Q47164","datavalue__string":null,"text":null,"language":null,"time":null,"timezone":null,"before":null,"after":null,"precision":null,"calendarmodel":null}},"rank":"normal"}],"P22":[{"mainsnak":{"property":"P22","datatype":"wikibase-item","datavalue":{"id":"Q3322144","datavalue__string":null,"text":null,"language":null,"time":null,"timezone":null,"before":null,"after":null,"precision":null,"calendarmodel":null}},"rank":"normal"}],"P166":[{"mainsnak":{"property":"P166","datatype":"wikibase-item","datavalue":{"id":"Q852071","datavalue__string":null,"text":null,"language":null,"time":null,"timezone":null,"before":null,"after":null,"precision":null,"calendarmodel":null}},"rank":"normal"}],"P410":[{"mainsnak":{"property":"P410","datatype":"wikibase-item","datavalue":{"id":"Q19100","datavalue__string":null,"text":null,"language":null,"time":null,"timezone":null,"before":null,"after":null,"precision":null,"calendarmodel":null}},"rank":"normal"}],"P26":[{"mainsnak":{"property":"P26","datatype":"wikibase-item","datavalue":{"id":"Q1095612","datavalue__string":null,"text":null,"language":null,"time":null,"timezone":null,"before":null,"after":null,"precision":null,"calendarmodel":null}},"rank":"normal"}],"P512":[{"mainsnak":{"property":"P512","datatype":"wikibase-item","datavalue":{"id":"Q1765120","datavalue__string":null,"text":null,"language":null,"time":null,"timezone":null,"before":null,"after":null,"precision":null,"calendarmodel":null}},"rank":"normal"}],"P106":[{"mainsnak":{"property":"P106","datatype":"wikibase-item","datavalue":{"id":"Q5446967","datavalue__string":null,"text":null,"language":null,"time":null,"timezone":null,"before":null,"after":null,"precision":null,"calendarmodel":null}},"rank":"normal"}],"P21":[{"mainsnak":{"property":"P21","datatype":"wikibase-item","datavalue":{"id":"Q6581097","datavalue__string":null,"text":null,"language":null,"time":null,"timezone":null,"before":null,"after":null,"precision":null,"calendarmodel":null}},"rank":"normal"}],"P646":[{"mainsnak":{"property":"P646","datatype":"external-id","datavalue":{"id":null,"datavalue__string":"/m/022jzf","text":null,"language":null,"time":null,"timezone":null,"before":null,"after":null,"precision":null,"calendarmodel":null}},"rank":"normal"}],"P18":[{"mainsnak":{"property":"P18","datatype":"commonsMedia","datavalue":{"id":null,"datavalue__string":"Kiefer Sutherland at 24 Redemption premiere 1 (headshot).jpg","text":null,"language":null,"time":null,"timezone":null,"before":null,"after":null,"precision":null,"calendarmodel":null}},"rank":"normal"}],"P1441":[{"mainsnak":{"property":"P1441","datatype":"wikibase-item","datavalue":{"id":"Q56194","datavalue__string":null,"text":null,"language":null,"time":null,"timezone":null,"before":null,"after":null,"precision":null,"calendarmodel":null}},"rank":"normal"}],"P1448":[{"mainsnak":{"property":"P1448","datatype":"monolingualtext","datavalue":{"id":null,"datavalue__string":null,"text":"Jack Bauer","language":"en","time":null,"timezone":null,"before":null,"after":null,"precision":null,"calendarmodel":null}},"rank":"normal"}],"P3373":[{"mainsnak":{"property":"P3373","datatype":"wikibase-item","datavalue":{"id":"Q10290844","datavalue__string":null,"text":null,"language":null,"time":null,"timezone":null,"before":null,"after":null,"precision":null,"calendarmodel":null}},"rank":"normal"}],"P3417":[{"mainsnak":{"property":"P3417","datatype":"external-id","datavalue":{"id":null,"datavalue__string":"Jack-Bauer","text":null,"language":null,"time":null,"timezone":null,"before":null,"after":null,"precision":null,"calendarmodel":null}},"rank":"normal"}],"P4839":[{"mainsnak":{"property":"P4839","datatype":"external-id","datavalue":{"id":null,"datavalue__string":"Entity[\"FictionalCharacter\", \"JackBauer\"]","text":null,"language":null,"time":null,"timezone":null,"before":null,"after":null,"precision":null,"calendarmodel":null}},"rank":"normal"}],"P175":[{"mainsnak":{"property":"P175","datatype":"wikibase-item","datavalue":{"id":"Q103946","datavalue__string":null,"text":null,"language":null,"time":null,"timezone":null,"before":null,"after":null,"precision":null,"calendarmodel":null}},"rank":"normal"}],"P1412":[{"mainsnak":{"property":"P1412","datatype":"wikibase-item","datavalue":{"id":"Q1860","datavalue__string":null,"text":null,"language":null,"time":null,"timezone":null,"before":null,"after":null,"precision":null,"calendarmodel":null}},"rank":"normal"}],"P1417":[{"mainsnak":{"property":"P1417","datatype":"external-id","datavalue":{"id":null,"datavalue__string":"topic/Jack-Bauer","text":null,"language":null,"time":null,"timezone":null,"before":null,"after":null,"precision":null,"calendarmodel":null}},"rank":"normal"}],"P5800":[{"mainsnak":{"property":"P5800","datatype":"wikibase-item","datavalue":{"id":"Q215972","datavalue__string":null,"text":null,"language":null,"time":null,"timezone":null,"before":null,"after":null,"precision":null,"calendarmodel":null}},"rank":"normal"}],"P6262":[{"mainsnak":{"property":"P6262","datatype":"external-id","datavalue":{"id":null,"datavalue__string":"24:Jack_Bauer","text":null,"language":null,"time":null,"timezone":null,"before":null,"after":null,"precision":null,"calendarmodel":null}},"rank":"normal"}],"P570":[{"mainsnak":{"property":"P570","datatype":"time","datavalue":{"id":null,"datavalue__string":null,"text":null,"language":null,"time":null,"timezone":null,"before":null,"after":null,"precision":null,"calendarmodel":null}},"rank":"normal"}],"P2581":[{"mainsnak":{"property":"P2581","datatype":"external-id","datavalue":{"id":null,"datavalue__string":"03111055n","text":null,"language":null,"time":null,"timezone":null,"before":null,"after":null,"precision":null,"calendarmodel":null}},"rank":"normal"}],"P10291":[{"mainsnak":{"property":"P10291","datatype":"external-id","datavalue":{"id":null,"datavalue__string":"12819","text":null,"language":null,"time":null,"timezone":null,"before":null,"after":null,"precision":null,"calendarmodel":null}},"rank":"normal"}],"P10757":[{"mainsnak":{"property":"P10757","datatype":"external-id","datavalue":{"id":null,"datavalue__string":"170","text":null,"language":null,"time":null,"timezone":null,"before":null,"after":null,"precision":null,"calendarmodel":null}},"rank":"normal"}],"P3553":[{"mainsnak":{"property":"P3553","datatype":"external-id","datavalue":{"id":null,"datavalue__string":"19959170","text":null,"language":null,"time":null,"timezone":null,"before":null,"after":null,"precision":null,"calendarmodel":null}},"rank":"normal"}]}}
